### ISSUE TITLE
Move file to new directory -> change package

### ISSUE THREAD

**Issue by mweibel** [go-lang-plugin-org/go-lang-idea-plugin#1711]
FYI I started the implementation of this feature. Renaming the package according to the new directory the file was moved in works in the meantime and I'm starting with the work on finding usages and renaming the import declarations there. 
Should I open a PR for the WIP I'm working on or what is the preferred process?

It's my first contribution to a intellij plugin, I expect you'll have quite some comments when reviewing ;-)

**Comment by ignatov** [comment]
Please feel free to open a PR. Sure we'll help you with comments.

**Comment by ignatov** [comment]
You can start from `com.intellij.refactoring.move.moveFilesOrDirectories.MoveFileHandler` and its implementations.

**Comment by mweibel** [comment]
Working on the linked branch, compare view: https://github.com/go-lang-plugin-org/go-lang-idea-plugin/compare/master...mweibel:feature/1711-move-file-handler?expand=1

Currently a bit struggling with how to set up the ReferencesSearch, it doesn't seem to find any references. Will need to look that up :)

**Comment by zolotov** [comment]
> Currently a bit struggling with how to set up the ReferencesSearch, it doesn't seem to find any references. 

Why do you think that there are references? Who added them?
How are you going to handle packages with _test suffix or `main` or `documentation` packages?
How are you going to handle following case?

/src/foo/foo.go
```
package foo_package
```
/src/bar/bar.go
```
package bar
```

`mv src/bar/bar.go src/foo/`

**Comment by dlsniper** [comment]
@mweibel please open a PR, it would be easier to comment there ;-) Also, great stuff, looking forward to it.

**Comment by zolotov** [comment]
> 1. get package, get all exported functions/types
> 2. search for usages of the package
> 3. for each found package usage, find which functions/types are used & add them to the map

No, it should be done by another 'rename package' refactoring

**Comment by mweibel** [comment]
@dlsniper created pr #1716 

@zolotov 

> Why do you think that there are references? Who added them?
sorry for being overly optimistic maybe ;) I thought the references are somehow added automatically by some "magic" ;) 

>  How are you going to handle packages with _test suffix or main or documentation packages?
Might need some special cases. 
I first try to get some basic stuff working because it's my very first contribution to any intellij plugin, never used the SDK before.

>  How are you going to handle following case?
Interesting case I honestly didn't think about yet. Most likely due to the following specification (even though this is a may and not a must):
" A set of files sharing the same PackageName form the implementation of a package. An implementation may require that all source files for a package inhabit the same directory. "

I'd probably need to have a look at the foo's package declaration (i.e. as a naive approach getting the "first" file in the target directory and it's package declaration) and use that one. 

This most likely means that my current `updateMovedFile()` method is flawed, but I'll take care of that as soon as I have the basic implementation done and more experience on how to do such things. 

> No, it should be done by another 'rename package' refactoring
So I should trigger a rename refactoring for each found usage basicly? 
I'd probably need something similar to `com.intellij.refactoring.util.MoveRenameUsageInfo`, no?

**Comment by zolotov** [comment]
> sorry for being overly optimistic maybe ;) I thought the references are somehow added automatically by some "magic" ;)

Indeed, before move file handler implementation it make sense to implement package references and `rename package` refactoring

> I first try to get some basic stuff working because it's my very first contribution to any intellij plugin, never used the SDK before.

I see. Maybe it's better to implement something simpler first (issue about packages refactoring is quite complicated actually, that's why it's still open). At least simpler issues will be merged faster :)  You can find relatively easy issues by `up for grabs` label.

> Due to the following specification (even though this is a may and not a must)...:
" A set of files sharing the same PackageName form the implementation of a package. An implementation may require that all source files for a package inhabit the same directory. "

Yes, my example is valid by spec and compilable by google compiler.

> .. I'd probably need to have a look at the foo's package declaration (i.e. as a naive approach getting the "first" file in the target directory and it's package declaration) and use that one.

Yes, see GoUtil#getAllPackagesInDirectory

> So I should trigger a rename refactoring for each found usage basicly? 

Not sure, but I guess that you should. 

> I'd probably need something similar to com.intellij.refactoring.util.MoveRenameUsageInfo, no?

Also not sure but looks like you need something similar. Anyway you should have package references before that.

**Comment by mweibel** [comment]
> Indeed, before move file handler implementation it make sense to implement package references and rename package refactoring

Ok I see. Will have a look at that but as you said, most likely something simple would be better to start with.

> Yes, see GoUtil#getAllPackagesInDirectory

Changed now to use that function instead of manually getting the files and it's first package :)

**Comment by kolkov** [comment]
What problem arose in the implementation of this functionality?
